File: conformance/tests/merge-locals-shapes/MergeLocalsShapes.runar.py
from runar import (
    StatefulSmartContract, Bigint, public,
)


class MergeLocalsShapes(StatefulSmartContract):
    """The three branch-merge arities that had no real-crypto evidence: k=2
    symmetric (both arms rebind both locals), k=3, and a nested `if` whose
    merge lands at the outer level. Companion to `branch-merged-locals`, which
    pins the asymmetric k=2 case."""

    a: Bigint
    b: Bigint
    c: Bigint

    def __init__(self, a: Bigint, b: Bigint, c: Bigint):
        super().__init__(a, b, c)
        self.a = a
        self.b = b
        self.c = c

    @public
    def both_arms(self, x: Bigint, flag: Bigint):
        na = self.a
        nb = self.b
        if flag > 0:
            na = x + 1
            nb = x + 2
        else:
            na = x + 3
            nb = x + 4
        self.add_output(1000, na, nb, self.c)

    @public
    def three(self, x: Bigint, flag: Bigint):
        na = self.a
        nb = self.b
        nc = self.c
        if flag > 0:
            na = x + 1
            nb = x + 2
            nc = x + 3
        else:
            na = x + 4
            nb = x + 5
            nc = x + 6
        self.add_output(1000, na, nb, nc)

    @public
    def nested(self, x: Bigint, outer: Bigint, inner: Bigint):
        na = self.a
        nb = self.b
        if outer > 0:
            if inner > 0:
                na = x + 1
            else:
                nb = x + 2
        self.add_output(1000, na, nb, self.c)

File: conformance/tests/merge-locals-shapes/MergeLocalsShapes.runar.java
package runar.conformance.mergelocalsshapes;

import runar.lang.StatefulSmartContract;
import runar.lang.annotations.Public;
import runar.lang.types.Bigint;

/**
 * MergeLocalsShapes -- the three branch-merge arities that had no real-crypto
 * evidence: k=2 symmetric (both arms rebind both locals), k=3, and a nested
 * {@code if} whose merge lands at the outer level. Companion to
 * {@code branch-merged-locals}, which pins the asymmetric k=2 case.
 */
class MergeLocalsShapes extends StatefulSmartContract {

    Bigint a;
    Bigint b;
    Bigint c;

    MergeLocalsShapes(Bigint a, Bigint b, Bigint c) {
        super(a, b, c);
        this.a = a;
        this.b = b;
        this.c = c;
    }

    @Public
    void bothArms(Bigint x, Bigint flag) {
        Bigint na = this.a;
        Bigint nb = this.b;
        if (flag.gt(Bigint.ZERO)) {
            na = x.plus(Bigint.ONE);
            nb = x.plus(Bigint.of(2));
        } else {
            na = x.plus(Bigint.of(3));
            nb = x.plus(Bigint.of(4));
        }
        this.addOutput(1000L, na, nb, this.c);
    }

    @Public
    void three(Bigint x, Bigint flag) {
        Bigint na = this.a;
        Bigint nb = this.b;
        Bigint nc = this.c;
        if (flag.gt(Bigint.ZERO)) {
            na = x.plus(Bigint.ONE);
            nb = x.plus(Bigint.of(2));
            nc = x.plus(Bigint.of(3));
        } else {
            na = x.plus(Bigint.of(4));
            nb = x.plus(Bigint.of(5));
            nc = x.plus(Bigint.of(6));
        }
        this.addOutput(1000L, na, nb, nc);
    }

    @Public
    void nested(Bigint x, Bigint outer, Bigint inner) {
        Bigint na = this.a;
        Bigint nb = this.b;
        if (outer.gt(Bigint.ZERO)) {
            if (inner.gt(Bigint.ZERO)) {
                na = x.plus(Bigint.ONE);
            } else {
                nb = x.plus(Bigint.of(2));
            }
        }
        this.addOutput(1000L, na, nb, this.c);
    }
}
